imap/http_caldav.c

```diff
@@ -1190,6 +1194,126 @@ static modseq_t caldav_get_modseq(struct mailbox *mailbox,
+
+    int r = annotatemore_lookupmask(calhomename, annotname, userid, &attrib);
+    if (!r && attrib.len) {
+        defaultname = buf_release(&attrib);
+    }
+    else defaultname = xstrdup(SCHED_DEFAULT);
+
+    size_t len = strlen(defaultname);
+    if (defaultname[len-1] == '/') defaultname[len-1] = '\0';
+
+    buf_free(&attrib);
+    free(calhomename);
+    return defaultname;
+}
+
+static int proppatch_scheddefault(xmlNodePtr prop, unsigned set,
+                                  struct proppatch_ctx *pctx,
+                                  struct propstat propstat[],
+                                  void *rock __attribute__((unused)))
+{
+    /* Only allow PROPPATCH on CALDAV:schedule-inbox-URL */
+    if ((pctx->txn->req_tgt.flags != TGT_SCHED_INBOX) || !set) {
+        xml_add_prop(HTTP_FORBIDDEN, pctx->ns[NS_DAV], &propstat[PROPSTAT_FORBID],
+                prop->name, prop->ns, NULL, DAV_PROT_PROP);
+        *pctx->ret = HTTP_FORBIDDEN;
+        return HTTP_FORBIDDEN;
+    }
+
+    /* Validate property value */
+    int precond = CALDAV_VALID_DEFAULT;
+    char *href = NULL;
+    mbname_t *mbname = NULL;
+
+    if (prop->children && !prop->children->next) {
+        xmlNodePtr cur = prop->children;
+        if (cur->type == XML_ELEMENT_NODE) {
+            if (!xmlStrcmp(cur->name, BAD_CAST "href")) {
+                href = (char*) xmlNodeGetContent(cur);
+                if (href && *href) {
+                    /* Strip trailing '/' character */
+                    size_t len = strlen(href);
+                    if (len > 1 && href[len-1] == '/') {
+                        href[len-1] = '\0';
+                    }
+                }
+            }
+        }
+    }
+
+    if (href) {
+        buf_reset(&pctx->buf);
+        if (strchr(httpd_userid, '@') || !httpd_extradomain) {
+            buf_printf(&pctx->buf, "%s/%s/%s/", namespace_calendar.prefix,
+                    USER_COLLECTION_PREFIX, httpd_userid);
+        }
+        else {
+            buf_printf(&pctx->buf, "%s/%s/%s@%s/", namespace_calendar.prefix,
+                    USER_COLLECTION_PREFIX, httpd_userid, httpd_extradomain);
+        }
+        if (!strncmp(href, buf_cstring(&pctx->buf), buf_len(&pctx->buf))) {
+            const char *cal = href + buf_len(&pctx->buf);
+            if (cal) {
+                char *mboxname = caldav_mboxname(httpd_userid, cal);
+                if (mboxname_iscalendarmailbox(mboxname, 0) &&
+                     mboxname_policycheck(mboxname) == 0) {
+                    mbname = mbname_from_intname(mboxname);
+                }
+                free(mboxname);
+            }
+        }
+    }
+
+    if (mbname) {
+        char *calhomename = caldav_mboxname(httpd_userid, NULL);
+        struct mailbox *calhome = NULL;
+        struct mailbox *mailbox = NULL;
+        int r = mailbox_open_iwl(calhomename, &calhome);
+        if (!r) r = mailbox_open_iwl(mbname_intname(mbname), &mailbox);
+        if (!r) {
```

Comment:
Why do we need to open the prospective default calendar if we're setting the annotation on cal-home-set?

Reply:
We don't want anyone deleting it before we are finished with the transaction.

Reply:
Yes, make sense.  I didn't fully form my previous thought.  Would a namelock be cheaper than opening and closing the mailbox?

Reply:
We need to lock the mailbox, the namelock is not enough.

Reply:
We need to lock the mailbox, a namelock is not enough.